[SLG-0003]: Standardized Error Metadata via Logger Convenience (#405)
apple/swift-log: Sources/Logging/Docs.docc/Proposals/SLG-0003.md

```

i.e. it's a convenience wrapper of the existing methods. We propose that such convenience methods are added for all log methods (even if it arguably makes less sense for e.g. `trace` and `debug` than `warning`). 

There are two clear gains from this: 

* It is obvious for the caller _how_ to attach an `Error` to a log (and the caller code also looks cleaner)
* We get a consistent format for logged errors

Of course, it might take time for this feature to be adopted, but the proposed solution is fully backwards compatible, meaning there is no added downside of having some parts of your code calling the new convenience methods, and other parts manually attaching the error.

### Detailed design

The proposed solution is to add a new overload for `Logger.log`, that accepts an `Error` as its third argument and serializes it, adding the type and message to the `metadata` parameter.

Additionally, we add a new overload for each log method (`trace`, ..., `critical`) that accepts an `Error` as its second argument, that in turn calls the new `log` method.

```swift

> public func log(

Reviewer: Oh, sorry for being unclear. I suggest adding the new methods **and** their docc comments, as this is also part of the "public interface" of the package.

> public func log(

Author: I see, added it now! Takes quite a bit of space with all the different methods 😅

        source: @autoclosure () -> String? = nil,
        file: String = #fileID,
        function: String = #function,
        line: UInt = #line
    ) {
        self.log(
            level: level, 
            message(), 
            metadata: { 
                var metadata = metadata()
                metadata["error.type"] = "\(String(reflecting: type(of: error)))"
                metadata["error.message"] = "\(error)"
                return metadata
            }(),
            source: source(), 
            file: file, 
            function: function, 
            line: line
        )
    }

public func critical(
        _ message: @autoclosure () -> Logger.Message,
        error: Error,
        metadata: @autoclosure () -> Logger.Metadata? = nil,
        source: @autoclosure () -> String? = nil,
        file: String = #fileID,
        function: String = #function,
        line: UInt = #line
    ) {
        self.log(
            level: .critical,
            message(),
            error: error,
            metadata: metadata(),
            source: source(),
            file: file,
            function: function,
            line: line
        )
    }
```

> We choose the keys `"error.message"` and `"error.type"` since they correspond to what [ECS](https://www.elastic.co/docs/reference/ecs/ecs-error) and [OpenTelemetry](https://opentelemetry.io/docs/specs/semconv/registry/attributes/error/#error-type) use.

Reviewer: Than you for adding. Could you also add a rejection reason for it?

However, it should be noted that users of `Logger` are still allowed to use the keys `"error"`, `"error.message"` and `"error.type"` manually. That is, both of the following are still valid.

```swift
logger.warning("Valid log", metadata: ["error": "\(error)"])
logger.warning("Valid log", metadata: ["error.message": ["msg1", "msg2"]])
```

This suggests that a `LogHandler` that wants to extract error information from the metadata must ensure that the expected types are passed. For example, it could opt to re-map any error metadata that contain only strings, and leave any other metadata using those keys untouched.

### API stability

The proposed additions are fully backwards compatible, meaning that all existing code will continue to function as is. Additionally, a user of `Logger` can start using the new convenience methods, without any modifications of the `LogHandler`, since a `LogHandler` is already expected to handle `Metadata`. 

> If an `Error` is passed using the new convenience methods, and the same keys (`"error.message"`, `"error.type"`) are added as metadata on the logger (or on a metadata provider), it is up to the `LogHandler` how to merge these metadata. This is not a new issue, as the same may already happen today, if the caller manually serializes the error using those keys. Really, the only potential issue is if a caller uses the new methods to pass an `Error`, while also passing metadata with the same keys.

Reviewer: ```suggestion
A potential metadata collision exists when an `Error` is passed via the new methods while the same keys (`"error.message"`, `"error.type"`) are present in metadata:
```

```swift
logger.warning("Something went wrong", error: error, metadata: ["error.message": "some other info"])
```

> We suggest that in this case, the `error` takes precedence, and the `"error.message"` from `metadata` is discarded. We don't expect this to be an issue in reality. 

Reviewer: ```suggestion
We recommend that the `error` parameter takes precedence in such cases, with conflicting metadata values discarded. This aligns with the principle that explicit `error:` parameters represent primary data, while the same keys in metadata are likely incidental. In practice, we don't expect callers to deliberately pass conflicting error information.
```

        _ message: @autoclosure () -> Logger.Message,
        error: Error,
        metadata: @autoclosure () -> Logger.Metadata? = nil,
        source: @autoclosure () -> String? = nil,
        file: String = #fileID,
        function: String = #function,
        line: UInt = #line
    )
}
```

A draft implementation can be found in [apple/swift-log#406](https://github.com/apple/swift-log/pull/406).

We choose the keys `"error.message"` and `"error.type"` since they correspond to what [ECS](https://www.elastic.co/docs/reference/ecs/ecs-error) and [OpenTelemetry](https://opentelemetry.io/docs/specs/semconv/registry/attributes/error/#error-type) use.

We propose that the metadata keys used (`"error.message"` and `"error.type"`) should be considered stable and part of the `swift-log` API. This makes it so that `LogHandler`s can reliably access this information. This is useful if a `LogHandler` wants to output the logs in a format which uses other keys for error information (e.g. re-map the error to `"err.message"` and `"err.type"`). To facilitate this, we suggest adding an enum with the metadata key names.

```swift
extension Logger {
    /// Metadata keys with specific use-cases

>     public enum WellKnownMetadataKey: String {

Reviewer: There is a good chance this enum will be extended with more keys in the future, so it should be up for backward compatible evolution. One option is [`@nonexhaustive`](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0487-extensible-enums.md), but this limits the supported Swift versions. Or a `struct` with `static` properties similar to [this](https://github.com/apple/swift-distributed-tracing/blob/main/Sources/InMemoryTracing/InMemoryTracer.swift#L197). I would lean towards the second option as `LogHandler` would not need to match over all the special keys supported by the `Logger`, but would rather reference a few interesting ones.

What do you think @czechboy0?
